Heiko Code Review

CodeLtDave opened the pull request.

CodeLtDave committed:
aktueller Stand kinda working

asda

CodeLtDave committed:
Revert "aktueller Stand kinda working"

This reverts commit 78cfb44d6532ee927103c8c98c154a441fe86af8.

CodeLtDave committed:
MakeStatic files init

Commit:
Almost working

Commit:
kleiner Schubser

CodeLtDave committed:
Working Rewrite but other errors

CodeLtDave committed:
Revert "Working Rewrite but other errors"

This reverts commit 9a56081415a75f1029d56aaea3674a292a739c18.

Commit:
right before change working

Commit:
working

Commit:
HEUREKA

CodeLtDave committed:
fix public modifier disappearing

leave the remeaining modifiers instead of replacing them

CodeLtDave committed:
beautify

CodeLtDave committed:
cleanup

CodeLtDave committed:
Revert "cleanup"

This reverts commit 8c021e42514c000b4628f2b5b16a949f11b9ed18.

Commit:
InvocationUpdate added but not working

CodeLtDave committed:
Chage Invocations, work in progress

Commit:
Working Invocation change in the same class

Commit:
Refactoring now working in different classes

MethodInvocations will now also be changed in different classes within the same project. In other projects its not working properly.

Commit:
Bugfix

Bugfix for Refactoring of more than one method in the same class that individually contain a Methodinvocation

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Commit:
Bugfix

Fixed Bug when refactoring a method with another return type than void. Now the Refactoring works for all return types.

Commit:
Changed MethodInvocations don't overwrite arguments anymore

Commit:
First MakeStaticRefactoring Test implemented

Commit:
First Test Working

CodeLtDave committed:
better code structure and cleanup

Commit:
More Tests added

Commit:
Additional Tests and Bugfix for Refactoring

If MethodDeclaration is modified before MethodInvocations, the Refactoring will fail

Commit:
ThrowException Test working

CodeLtDave committed:
added ChangeManager which handels adding edits

Before there could be multiple CompilationUnitChanges for one CompilationUnit. Now you can use

addEditToChangeManager(TextEdit, ICompilationUnit)

so that this is automatically take care off. If there is a CompilationUnitChange already its pulled from ChangeManager and edited instead of having multiple CompilationUnitChanges on one ICompilationUnit.

CodeLtDave committed:
cleanup

CodeLtDave committed:
small fix

CodeLtDave committed:
small refactorings

Commit:
new tests

new tests for every type as parameter and return type

Commit:
new test for this-usage in MethodDeclaration

Commit:
Additional Tests for multiple refactored files

CodeLtDave committed:
Improve method invocation handling in MakeStaticRefactoring

Enhance the MakeStaticRefactoring class to better handle method
invocations in both instance and static contexts. Adjustments
include passing 'this' as the new argument for invocations within
instance methods and passing the instance variable for invocations
in other classes or static methods.

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Commit:
small change in test

Commit:
Additional test for this + InnerClass

Commit:
small fix for test

CodeLtDave committed:
Implement handling of instance usages in method refactoring now with a CHECK

This is not finished yet, there are 2 tests failing but at least no errors anymore

In this commit, we introduced the fHasInstanceUsages flag to track if a method being refactored has any instance variables or methods. If fHasInstanceUsages is true, the method will be modified to accept an additional parameter representing the instance of the class.

Key changes include:

Implementing an ASTVisitor to visit SimpleName nodes within the method body and check for instance variables or methods
If an instance variable is found, the code replaces the variable with an appropriate FieldAccess expression
If an instance method is found, the code replaces the method invocation with an appropriate MethodInvocation expression

This update improves the refactoring process by correctly handling methods with instance usages, enabling a more accurate static method conversion.

Commit:
Beautify tests and proper suggestion of refactoring

Based on the specific selection of the user the refactoring will now only be suggested if the user selects a method.

Commit:
Additional functionality and adjustments

Additional checks for Refactoring in checkInitialConditions. Editor Selection for Refactoring is now more precisely.
Adjustments for tests for more accurate selection.
Refactoring can now be executed from outline window.
Message Constants included for better Exception and RefactoringStatus handling.

Commit:
Adjustments to tests

Commit:
New Tests and handling for SuperMethodInvocations

Commit:
Little fix for tests that expect error

Commit:
Implemented ASTMatcher for Tests

The method matchASTs is more precise and ignores eventual whitespaces in the files. With the method matchFiles it is easier to understand where the files have differences.  Both methods can be used for testing

Commit:
Check if parameter with same name already exists

if Method has instance usage a new parameter is introduced to the method. If this parameter has the same name as an already existing parameter then an error will be thrown.

Commit:
Check if duplicate method signature exists

when refactoring a method with instance usage, a new parameter is introduced. After the refactoring the method could have the same signature as an already existing method. This check throws an Error in this case.

Commit:
Bugfix Overlapping TextEdits when status has error

Commit:
Multiple Updates

- Better Handling of SuperMethodInvocations.
- Deletion of @Override Annotation.
- Better Handling of StatusErrors -> immediate return after error

Commit:
ShortCut implemented: Alt+Shift+K

Commit:
Bugfix for box name on Refactoring errors

Commit:
MakeStatic can now be used from upper Refactoring menu

Commit:
Cleanup

Commit:
Check if selected method is already static + Bugfix

- new test for this case
- new error message
- fixed a bug where an "Overlapping text edits" error was thrown when pressing back in WIzard after an error occured and then again pressing continue to continue the refactoring.
- Changed a line, so that MakeStatic error message and not IntroduceIndirection error  message will be thrown in MakeStaticAction

Commit:
Adjusted Handling of SuperMethodInvocations and methods that are overriden in TypeHierarchy

Commit:
Code Cleanup

- deleted method getCachedCURewrite()
- more comments for better understanding
- fMethodDeclaration is not initialized in checkInitialConditions() and not in findMethodDeclaration()
- findMethodDeclaration() is now called checkForInstanceUsagesInMethodDeclaration() and is only for this use.

Commit:
New Tests for special cases

- Generic declaration
- Non Static Inner Class

CodeLtDave committed:
tests working

CodeLtDave committed:
Merge branch '_DavidNEW' into _MichaelNEW

Commit:
Additional tests and handling

Tests for following cases:
- Usage of Generics
- refactoring a method should not add additional parameter in case of field access whenparameter already exists
- Selected method is overriding a method of parent type and has no parameter -> refactoring would change method so that it hides parent type

Also included handling for last case

Commit:
Additional Tests

new tests for following cases:
- Field access in anonymous class of a field that belongs to the outer class
- The same case but in a lambda and not in an anonymous class

CodeLtDave committed:
removed findMethodDeclaration and added functionality to new inner class"ChangeInstanceUsagesInMethodBody"

Commit:
New Tests

- Many new tests for various use cases

Commit:
Better handling of duplicate method signature

Before method were only compared with number of parameters and  method name. Now also the parameter types are compared

Commit:
Better handling for duplicate parameterName

Commit:
Handling for static method in lower hierarchy

- handling added if method with same signature in lower hierarchy is static -> refactoring only throws error if method in lower hierarchy overrides selected method, but a static method can not override.
- Bug fixed for test

Commit:
Adjustments to tests, Little change to input page, Bugifx

- Fixed a bug where methods that override parent method and have instance usage couldnt be refactored
- Adjustments to tests so they work properly
- Deleted not needed tests
- Little Change to Input Page and Wizard

Commit:
Added warning when refactoring method that overrides

- Changed dialog window size back to default

Commit:
implemented check for recursive method

- User gets error message when refactoring a recursive method.
- This behavior does not apply for SuperMethodinvocations. Methods that override the parent method and call it with super can still be refactored.

Commit:
Added functionality for updating javadocs

Commit:
Adjustment to ExtraTests

Commit:
Added functionality for Generics

- Adjustments to tests
- methods with generics can now be refactored but only if they are not already parametrized with type parameters

Commit:
Updated Generic functionality

- methods that use less generic parameters than the amount of type parameters of the class will now be refactored correctly. Only the exact amount of needed parameters will be added to the type parameter list of the method.
- New Tests for special cases with generics added

Commit:
bug fix if method has array as parameter

Commit:
New Test

Commit:
Several bugfixes for generics functionality

- additional test

Commit:
new Generics functionality

- added functionality for already parametrized methods
- added functionality for bounds of typeParameters

Commit:
little fix for test

- parametrized methods can be refactored now, so test doesnt need to check for error

Commit:
check for WildCardTypes as bounds implemented

- additional Tests and check for wildCardTypes as bounds of typeparam of class. (T extends List<?>)

Commit:
Bugfix and adjustment to tests

- Refactoring is now possible if a method does not override the parent type method but has the same name as the parent type method

Commit:
Bugfix for generics in JavaDoc

- now braces <T> are also added to JavaDoc when adding type parameters

Commit:
check for LambdaExpression and MethodReference

- if Lambda is detected as MethodInvocation the refactoring is cancelled
- bugfix for invoking refactoring on MethodInvocation that is in different class than MethodDeclaration
- code cleanup

Commit:
Functionality for VarArgs included

- new parameter for instance access is now inserted as first parameter of method, so that varargs are always the last parameter
- changed every test that adds new parameter to new behavior

Commit:
Cleanup

- extracted methods from method modifyMethodDeclaration() for more clarity

Commit:
little change

CodeLtDave committed:
fix: hasInstanceUsage and InnerClassCreatation

fix for wrongfully recognised hasInstanceUsages and creating anonymous inner instance classes

Commit:
Bugfix for generics

- updateTypeParamList now also includes typeParams of parametrized methodParams

Commit:
small changes to some tests

Commit:
changes to Tests

CodeLtDave committed:
unfinished state before extracting handleInstanceVariable

Commit:
merging of test classes and changes for MethodReferences

- Changed behaviour for MethodReferences and SuperMethodReferences
- LambdaExpressions are now allowed for Refactoring
- Merging of test classes

Commit:
New tests

- It is not possible anymore to refactor a method that uses explicit super method invocations in the methodDeclaration body
- New warning for explicit access of fields from the parent type with the super keyword
-

Commit:
Comparing of methods is now more precise

Commit:
new test for foncatenated field access

CodeLtDave committed:
extend testConcatenatedFieldAccessAndQualifiedNames

CodeLtDave committed:
fixed testConcatenatedFieldAccessAndQualifiedNames

CodeLtDave committed:
fixes for modifyMethodInvocation and concatinated FieldAccess/QualifiedName

CodeLtDave committed:
fixes for modifyMethodInvocation

CodeLtDave committed:
testsWorking, fixed behavior for InnerFieldAccessInAnonymousClass

Commit:
Deleted output files for tests with error...

-  check if source is available for selected method
- New test for this case

CodeLtDave committed:
checks for anonymous/inner classes

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Commit:
right output files deleted and others inserted again

CodeLtDave committed:
fixed test22?

CodeLtDave committed:
Letzter SCHUBSSER

Commit:
Merging of tests and new test for nested classes

Commit:
Test merging completed

CodeLtDave committed:
fixed testNested

CodeLtDave committed:
comment refactorings to modifyMethodDeclaration

CodeLtDave committed:
refactorings, broke testInheritance3

CodeLtDave committed:
structure fix to checkNode

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/refactoring.properties:
```diff
@@ -1343,6 +1343,27 @@ IntroduceIndirectionRefactoring_call_warning_anonymous_cannot_qualify=The declar
+MakeStaticRefactoring_not_available_for_local_or_anonymous_types=Make Static is not supported on methods inside local or anonymous types.
+MakeStaticRefactoring_not_available_for_constructors=Make Static is not available for constructors.
+MakeStaticRefactoring_checking_activation=Checking activation of Make Static...
+MakeStaticRefactoring_not_available_for_super_method_invocations=Make Static is not available for SuperMethodInvocations.
```
Write "super method invocations" instead of "SuperMethodInvocation" since the latter is a technical/internal name that is usually unknown to the end user.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/refactoring.properties:
```diff
@@ -1343,6 +1343,27 @@ IntroduceIndirectionRefactoring_call_warning_anonymous_cannot_qualify=The declar
+MakeStaticRefactoring_hiding_method_of_parent_type=Make Static is not available. Method of parent type would be hidden after refactoring.
+MakeStaticRefactoring_not_available_for_recursive_methods=Make Static is not available for recursive methods.
+MakeStaticRefactoring_not_available_for_parametrized_methods=Make Static is not available for parametrized methods.
+MakeStaticRefactoring_not_available_for_wildCardTypes_as_bound=Make Static is not available if class has WildCardTypes as bounds of its type parameters.
```
Write "has wildcards as bounds" instead of "has WildCardTypes as bounds" since the latter uses a technical/internal name that is usually unknown to the end user.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/refactoring.properties:
```diff
@@ -1343,6 +1343,27 @@ IntroduceIndirectionRefactoring_call_warning_anonymous_cannot_qualify=The declar
+MakeStaticRefactoring_not_available_for_parametrized_methods=Make Static is not available for parametrized methods.
+MakeStaticRefactoring_not_available_for_wildCardTypes_as_bound=Make Static is not available if class has WildCardTypes as bounds of its type parameters.
+MakeStaticRefactoring_not_available_for_method_references=Make Static is not available if method references use selected method.
+MakeStaticRefactoring_explicit_super_method_invocation=The method contains explicit invocation of a superclass method using 'super' and will lead to behavior changes and break the class hierarchy.
+MakeStaticRefactoring_selected_method_uses_super_field_access=The method contains explicit access of a superclass field using 'super' and will lead to behavior changes and break the class hierarchy.
```
You could make two sentences out of each of these to improve readabiliy and improve grammatics (currently you say "the method [...] will lead to behavior changes" ;-) )

After 'super', you can let the first sentence end and start another with "Applying the refactoring will lead [...]".

Question: "Will" this always lead to changes of the program behavior and break the class hierarchy or should it actually be "can"?

Michael5601 replied:
It now says "Applying the refactoring would lead to behavior changes". The class hierarchy would not be broken by performing the refactoring. But the semantics would change because an instance of the subclass would be used instead of the super keyword. So it is right to block this case with an error message but the message is now less confusing.

HeikoKlare commented on org.eclipse.jdt.ui.junit.sampleproject/.project:
```diff
@@ -11,13 +11,13 @@
 		<buildCommand>
-			<name>org.eclipse.m2e.core.maven2Builder</name>
+<!-- 			<name>org.eclipse.m2e.core.maven2Builder</name> -->
 			<arguments>
 			</arguments>
 		</buildCommand>
 	</buildSpec>
 	<natures>
-		<nature>org.eclipse.m2e.core.maven2Nature</nature>
+<!-- 		<nature>org.eclipse.m2e.core.maven2Nature</nature> -->
```
Undo these changes.

CodeLtDave replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/resources/MakeStatic/testNoAdditionalParameter2/in/Foo.java:
```diff
@@ -0,0 +1,9 @@
+public class Foo {
+
+	String j= "";
+
+	public String bar(Foo foo) {
+		String i= foo.j;
+		return i;
+	}
+}
```
Is there a difference to the test "testNoAdditionalParameter"?

Michael5601 replied:
No this is exactly the same. Back then the behaviour of the refactoring was a little bit different and the second test case had different parameter names in the method so both cases were important. After I uniformed all names to "foo" and "bar" the test case was obsolete. I deleted it now.

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/resources/MakeStatic/testSourceNotAvailable/in/Foo.java:
```diff
@@ -0,0 +1,5 @@
+public class Foo {
+	public void bar() {
+		System.out.println();
+	}
+}
```
Just for my understanding: Why is this case unsupported? Or more precisely: What exactly is the "non-available source" here?

Michael5601 replied:
The refactored method in this case is System.out.println and not the method bar. Before the refactoring was causing issues when refactoring a method which source code is not reachable. We implemented the detection of this case and the refactoring now shows an error message to the user.

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/MakeStaticRefactoringTests.java:
```diff
@@ -0,0 +1,557 @@
+	private RefactoringStatus helper(String[] topLevelName, int startLine, int startColumn, int endLine, int endColumn)
+			throws Exception, JavaModelException, CoreException, IOException {
+
+		ICompilationUnit[] cu= new ICompilationUnit[topLevelName.length];
+		for (int i= 0; i < topLevelName.length; i++) {
+			String packName= topLevelName[i].substring(0, topLevelName[i].indexOf('.'));
```
Please you speaking names to improve readability of the code:
- "compilationUnits" instead of "cu"
- "packageName" instead of "packName"
- some reasonable name for the loop variable instead of "i", or even better use a for-each loop, if possible.

And please also check the rest of the code whether names can be improved.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/MakeStaticRefactoringTests.java:
```diff
@@ -0,0 +1,557 @@
+
+		ISourceRange selection= TextRangeUtil.getSelection(cu[0], startLine, startColumn, endLine, endColumn);
+
+		MakeStaticRefactoring ref= new MakeStaticRefactoring(cu[0], selection.getOffset(), selection.getLength());
```
Please use "refactoring" instead of "ref", not only because it is easer to read but also because "ref" is usually short for "reference" and can be misleading here.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/MakeStaticRefactoringTests.java:
```diff
@@ -0,0 +1,557 @@
+		return REFACTORING_PATH;
+	}
+
+	private RefactoringStatus helper(String[] topLevelName, int startLine, int startColumn, int endLine, int endColumn)
```
Please rename this method with a speaking name reflecting what it does. Also rename the first parameter: it sounds like it contains a name separated into segments, but it seems like it is simply a list of names. So first, it should be called "names" rather than "name", and you can also improve it by renaming it so it is clear what a "topLevelName" refers to.

Michael5601 replied:
The first parameter is now called "qualifiedClassNames" and the method is now called "performRefactoringAndMatchFiles".

HeikoKlare commented on org.eclipse.jdt.ui/META-INF/MANIFEST.MF:
```diff
@@ -135,5 +136,7 @@ Require-Bundle:
- org.eclipse.equinox.bidi;bundle-version="[0.10.0,2.0.0)"
+ org.eclipse.equinox.bidi;bundle-version="[0.10.0,2.0.0)",
+ org.junit,
+ junit-jupiter-api
```
Are these additional dependencies and exports necessary?

CodeLtDave replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/plugin.properties:
```diff
@@ -340,6 +340,8 @@ Refactoring.extractMethodAction.label=E&xtract Method...
 Refactoring.extractTempAction.label=Extract &Local Variable...
 Refactoring.extractConstantAction.label=Extr&act Constant...
 
+Refactoring.makeStaticAction.label=Make Static...
```
You could mark a symbol that is used for accessing the action via keypress. For example the "k" seems to be unused in that menu.

CodeLtDave replied:
added mark for symbol, but doesnt seem to work properly

HeikoKlare commented on org.eclipse.jdt.ui/ui refactoring/org/eclipse/jdt/internal/ui/refactoring/refactoringui.properties:
```diff
@@ -538,6 +538,15 @@ IntroduceIndirectionInputPage_update_references=&Redirect all method invocations
+#######################################
+# Make Static
+#######################################
+MakeStaticAction_title=Make Static...
```
Like mentioned above for the toolbar menu entry:
You could mark a symbol that is used for accessing the action via keypress. For example the "k" seems to be unused in that menu.

CodeLtDave replied:
Done, but doesn't seem to work.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticWizard.java:
```diff
@@ -0,0 +1,54 @@
+
+	public MakeStaticWizard(MakeStaticRefactoring ref){
+		super(ref, DIALOG_BASED_USER_INTERFACE | PREVIEW_EXPAND_FIRST_NODE);
+		setDefaultPageTitle("Make method static"); //$NON-NLS-1$
```
Please extract this string into a localizable variable (`*.properties` file like the other localizable strings).

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticWizard.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class MakeStaticWizard extends RefactoringWizard {
+
+	/* package */ static final String DIALOG_SETTING_SECTION= "MakeStaticWizard"; //$NON-NLS-1$
```
This variable seems to be unused? So please remove.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticWizard.java:
```diff
@@ -0,0 +1,54 @@
+		setDialogSettings(JavaPlugin.getDefault().getDialogSettings());
+	}
+
+	public Change createChange(){
+		// creating the change is cheap. So we don't need to show progress.
+		try {
+			return getRefactoring().createChange(new NullProgressMonitor());
+		} catch (CoreException e) {
+			JavaPlugin.log(e);
+			return null;
+		}
+	}
```
This method seems to be unused?

Michael5601 replied:
works fine without the method. I got this from another refactoring. It is deleted now.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticInputPage.java:
```diff
@@ -0,0 +1,56 @@
+	private final MakeStaticRefactoring fRefactoring;
+
+	public MakeStaticInputPage(MakeStaticRefactoring refactoring) {
+		super("MyMakeStaticRefactoringInputPage"); //$NON-NLS-1$
+		fRefactoring= refactoring;
+    }
+
+    @Override
+	public void createControl(Composite parent) {
+    	  // Create the SWT controls for your InputPage
+    	Composite composite= new Composite(parent, SWT.NONE);
+		composite.setLayout(new GridLayout(2, false));
+		GridData gridData = new GridData(SWT.FILL, SWT.FILL, true, true);
+	    composite.setLayoutData(gridData);
+		composite.setFont(parent.getFont());
+
+
+		Label label1= new Label(composite, SWT.NONE);
+		label1.setText("Make selected method static"); //$NON-NLS-1$
+		label1.setLayoutData(new GridData());
+		setControl(composite);
+
+		Dialog.applyDialogFont(composite);
+
+    }
+
+
+
+
+    @Override
+	protected boolean performFinish() {
+		return super.performFinish();
+	}
+
+	@Override
+	public IWizardPage getNextPage() {
+		return super.getNextPage();
+	}
+
+}
```
I am a little curious about this wizard page. Seems to me like it's implementation has not been finished? E.g., the passed refactoring is unused, Strings are defined within the code, variables have  default names, overwritten methods are obsolete etc.

Michael5601 replied:
I was trying around and forgot to delete some things. The Strings are now localized, unused variables deleted, variables have better names and obsolete methods are deleted.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticWizard.java:
```diff
@@ -0,0 +1,54 @@
+ * @since 3.29
+ *
+ */
+public class MakeStaticWizard extends RefactoringWizard {
```
Please add some Javadoc for class and public methods (i.e., the constructor).

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+
+import org.eclipse.jface.viewers.IStructuredSelection;
+
+/*******************************************************************************
+ * Copyright (c) 2000, 2011 IBM Corporation and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
```
Replace this with a new, appropriate copyright notice at the top of the file.

CodeLtDave replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+import org.eclipse.jdt.internal.ui.refactoring.RefactoringMessages;
+import org.eclipse.jdt.internal.ui.util.ExceptionHandler;
+
+/**
+ * Extracts an expression into a constant field and replaces all occurrences of the expression with
+ * the new constant.
```
This does not seem to be appropriate :-)

Michael5601 replied:
Changed it to "Modifies a method with the static keyword and changes all method invocations to properly invoke the refactored method."

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+ * Extracts an expression into a constant field and replaces all occurrences of the expression with
+ * the new constant.
+ *
+ * <p>
+ * This class may be instantiated; it is not intended to be subclassed.
+ * </p>
```
If it should not be sublassed, make it `final` and remove this comment.

Michael5601 replied:
I made it final and removed the comment.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+	/**
+	 * Note: This constructor is for internal use only. Clients should not call this constructor.
+	 *
+	 * @param editor the java editor
```
Document whether the editor is allowed to be `null` here.

CodeLtDave replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+	// Selected in Editor window
+	@Override
+	public void selectionChanged(ITextSelection selection) {
+		setEnabled(fEditor != null && SelectionConverter.getInputAsCompilationUnit(fEditor) != null);
```
Is it valid that the refactoring is initialized with a `null` editor? And in that case, can this method be called at all, as not editor exists?

Michael5601 replied:
I deleted the part "fEditor != null" because there is no way to call this method when the editor is null as far as I know.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+		}
+	}
+
+	/**
+	 * Note: This method is for internal use only. Clients should not call this method.
+	 *
+	 * @param selection the Java text selection (internal type)
+	 *
+	 * @noreference This method is not intended to be referenced by clients.
+	 */
```
You can refer to the documentation in the super class, as this is only a repetition?

CodeLtDave replied:
Super class has a diffrent documentation and doesn't look repetetive:

```
/**
 * Action that dispatches the <code>IAction#run()</code> and the
 * <code>ISelectionChangedListener#selectionChanged</code>
 * according to the type of the selection.
 *
 * <ul>
 * 	<li>if selection is of type <code>ITextSelection</code> then
 * 	<code>run(ITextSelection)</code> and <code>selectionChanged(ITextSelection)</code>
 * 	is called.</li>
 * 	<li>if selection is of type <code>IStructuredSelection</code> then
 * 	<code>run(IStructuredSelection)</code> and <code>
 * 	selectionChanged(IStructuredSelection)</code> is called.</li>
 * 	<li>default is to call <code>run(ISelection)</code> and <code>
 * 	selectionChanged(ISelection)</code>.</li>
 * </ul>
 *
 * <p>
 * Note: This class is not intended to be subclassed outside the JDT UI plug-in.
 * </p>
 *
 * @since 2.0
 *
 * @noextend This class is not intended to be subclassed by clients.
 */
```

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticAction.java:
```diff
@@ -0,0 +1,151 @@
+		try {
+			setEnabled(RefactoringAvailabilityTester.isMakeStaticAvailable(selection));
+		} catch (JavaModelException e) {
+			if (JavaModelUtil.isExceptionToBeLogged(e))
+				JavaPlugin.log(e);
```
Please put "then" statements into a block, even if it is only one (also applies to the rest of the class).

Michael5601 replied:
Done.

HeikoKlare reviewed (commented):
This is a review for the whole contribution except the `MakeStaticRefactoring` class. I will provide a separate review for that class.

In general, the part of the contribution I have looked at looks really good 👍
My comments are primarily minor. General comments not only dedicated to specific places in the code are the following:
* Improve names of methods and variables. I have given some pointer but please go through the code and look for further possibilities for improvements, in particular in cases where code is similar to / adapted from existing refactorings.
* Move the `MakeStatic...` classes to the appropriate packages. Currently they are placed within a package that is dedicated to the proposed refactoring.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+
+	//Convert ICompialtionUnit to CompilationUnit
+	private CompilationUnit convertICUtoCU(ICompilationUnit compilationUnit) {
+		ASTParser parser= ASTParser.newParser(AST.JLS20);
```
You should not fix the used Java version but the one currently used for the AST within the workspace/project. It seems to be provided by the shared variable `IASTSharedValues.SHARED_AST_LEVEL`.

CodeLtDave replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+ *
+ * @since 3.29
+ *
+ */
```
Please add a Javadoc describing what the refactoring does.

It would also be useful to have a short documentation of the fields describing their purpose, possible values etc.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+	private IMethodBinding fTargetMethodBinding;
+
+	public MakeStaticRefactoring(ICompilationUnit inputAsCompilationUnit, int offset, int length) {
+		fSelectionStart= offset;
+		fSelectionLength= length;
+		fSelectionCompilationUnit= inputAsCompilationUnit;
```
Please make these fields final so that no one overwrites them by accident.

CodeLtDave replied:
These fields can only be set in one of the two constructors. Making them final leads to compilation errors.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+
+	private IMethodBinding fTargetMethodBinding;
+
+	public MakeStaticRefactoring(ICompilationUnit inputAsCompilationUnit, int offset, int length) {
```
Please document the constructors and, in particular, the meaning of their parameters.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+
+		IMethod method= org.eclipse.jdt.internal.corext.refactoring.Checks.findMethod(methodName, parameterAmount, false, type);
+
+		if (method != null) {
```
In cases where such a check realizes a precondition for executing the method, better do the shortcut `if (method == null) { return; }`. It reduces the depth of the following code and makes it easier to follow.

CodeLtDave replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+		}
+	}
+
+	private final class ChangeInstanceUsagesInMethodBody extends ASTVisitor {
```
Instead of having this class modify the outer class (in particular the `fHasInstanceUsages` field), better store the result in a field of this inner class and make the `modifyMethodDeclaration` write that field back to the outer class after applying this visitor. Then you can also move this inner class to a standalone class, which one can read and comprehend in isolation.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+		@Override
+		public boolean visit(SimpleName node) {
+			IBinding binding= node.resolveBinding();
+			if (binding instanceof IVariableBinding) {
```
This method is rather long and complex to read. Better split it up by extracting parts of its implementation into separate methods. For example, the contents of both top-level if/else blocks (for IVariableBinding and IMethodBinding) can be moved into seprate methods. May the contents of the inner if blocks can then also be extracted into separate methods.
Such separations make the code easier to read, as it decomposes the functionality into smaller chunks. You can give each chunk a name, as it is wrapped into a method, that reflects what this code chunk is supposed to do.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+			else
+				node= checkNode(NodeFinder.perform(root, offset, length));
+		} catch (JavaModelException e) {
+			//TODO no catch
```
Is this exception supposed to be ignored?

Michael5601 replied:
done.

HeikoKlare reviewed (commented):
Before going into a detailed review for the `MakeStaticRefactoring` class, I have a general proposal for improving that class, which should be processed before having an in-depth review. The detailed comments I have already provided give some hints with respect to improving code style (clean code), to reduce the complexity of the code and make it easier to read and comprehend. This will make the refactoring look different compared to the existing ones, but will lead to code of higher quality.

Currently, the class is quite large. It mixes up different starting points (text selection in compilation unit vs. method in hand), which merge to a common control flow after checking initial conditions (which calculates the method out of the compilation unit + offset + length or vice versa). In addition, several methods do more than they claim to do (e.g., the methods for checking initial conditions do not only do a check but also change the state of the object). Just as an information: It is, in general, an indicator for good design if query functionality (retrieving information, validating etc.) is clearly separated from modification functionality (changing the state of an object).

I propose that you try the following to improve code quality:
Extract a `MakeStaticRefactoringContext` class out of the existing `MakeStaticRefactoring` class. This context class stores the compilation unit, the method, the offset and the length. The first two are accessible via getters to be used by the refactoring class. This context class is instantiated by the `MakeStaticRefactoring` class on its instantiation and can be created with either a compilation unit or with a method (just like the `MakeStaticRefactoring` is now). This context class also provides a method such as `calculateCompleteContext`, which calculates the method out of the compilation unit with offset and length or the compilation unit out of the method, like done during the initial condition check right now.
When the `MakeStaticRefactoring` checks the initial conditions, it calls this method to complete the context. This modification clearly separates the functionality to calculate the complete context (CU + method) from the refactoring itself. 
You can try to create one class for each of the steps (checking initial conditions, checking final conditions and creating the change), so that the central refactoring class only wraps these classes and executes the individual steps. This will even further improve comprehensibility by separating the logic into smaller parts. I would propose that you place the `MakeStaticRefactoring` with all these associated classes, such as the context class and potentially further classes, into a subpackage (e.g. `org.eclipse.jdt.internal.corext.refactoring.code.makestatic`).

Maybe you try this larger refactoring on a separate branch derived from this one to ensure that you can easily rollback if something goes wrong.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,885 @@
+	private void deleteOverrideAnnotation(ASTRewrite rewrite) {
+		ListRewrite listRewrite= rewrite.getListRewrite(fTargetMethodDeclaration, MethodDeclaration.MODIFIERS2_PROPERTY);
+		for (Object obj : fTargetMethodDeclaration.modifiers()) {
+			if (obj instanceof org.eclipse.jdt.core.dom.MarkerAnnotation) {
```
Forgot to mention this one in the review: In current Java, you can write:
```java
if (obj instanceof org.eclipse.jdt.core.dom.MarkerAnnotation marker) {
   // Use marker typed as MarkerAnnotation here
}
```

Michael5601 replied:
Done.

Commit:
fixed bug for testInheritance3

Commit:
error messages now better to understand

Commit:
deleted obsolete test testNoAdditionalParameter2

Commit:
Put all "then" statements into a block for readability. Autoformated code

CodeLtDave committed:
Revert accidental change in file

Commit:
improved readability for test class

CodeLtDave committed:
Move MakeStaticAction to org.eclipse.jdt.ui\ui\org\eclipse\jdt\ui\actions

CodeLtDave committed:
Move InputPage and Wizard

CodeLtDave committed:
Move MakeStaticRefactoring.java

CodeLtDave committed:
Revert "Move MakeStaticRefactoring.java"

This reverts commit 6005e5706ab3510f5f5dcc03988d85f3a1fe779c.

Commit:
changes for better understanding

- changed dialog title
- outsourced strings to localized variable

CodeLtDave committed:
Move MakeStaticRefactoring.java

CodeLtDave committed:
Merge branch '_DavidNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _DavidNEW

CodeLtDave committed:
MakeStaticWizard removed unused imports

CodeLtDave committed:
Added "&k" symbol for acessing the action via keypress

CodeLtDave committed:
adjusted copyright notice MakeStaticAction

CodeLtDave committed:
Merge branch '_DavidNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _DavidNEW

CodeLtDave committed:
Modified all Copyright notices

CodeLtDave committed:
changed AST version for parser to IASTSharedValues.SHARED_AST_LEVEL

CodeLtDave committed:
java docs edit, "if...return" restructure

Commit:
Improved code readability for inner class ChangeInstanceUsagesInMethodBody

Commit:
changed inner class ChangeInstanceUsagesInMethodBody to a standalone class

Commit:
deleted unnecessary line of code

Commit:
CleanUp MakeStaticInputPage

Commit:
new names for variables in InputPage

Commit:
small fix to Exception and deleted unnecessary null check

Commit:
JavaDocs for MakeStaticWizard

Commit:
name change for ChangeInstanceUsageInMethodBody

CodeLtDave committed:
create package for MakeStaticRefactoring.java, class split preperation

Commit:
JavaDocs for InstanceUsageRewriter

Commit:
JavaDiocs for MakeStaticRefactoring

Commit:
JavaDocs for fields in MakeStaticRefactoring

Commit:
Outsourced inner class MethodReferenceFinder and simplified its methods

Commit:
JavaDocs for MakeStaticRefactoring methods

CodeLtDave committed:
init ContextCalculator

Commit:
Merged methods isOverridden and is Overriding

Commit:
added field fStatus and changed all usages of status

Commit:
Copyright comment for classes

CodeLtDave committed:
progress

Commit:
improved readability for method modifyMethodInvocation

Commit:
tests for javaDocs

CodeLtDave committed:
unfinished

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/MakeStaticRefactoringTests.java:
```diff
@@ -0,0 +1,569 @@
+	private static final String REFACTORING_PATH= "MakeStatic/";
+
+	public MakeStaticRefactoringTests() {
+		rts= new RefactoringTestSetup();
```
Please use `this(rts)` instead of setting fields of super class.

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/MakeStaticRefactoringTests.java:
```diff
@@ -0,0 +1,569 @@
+			throws Exception, JavaModelException, CoreException, IOException {
+
+		List<ICompilationUnit> compilationUnits= new ArrayList<>();
+		//ICompilationUnit[] compilationUnits= new ICompilationUnit[qualifiedClassNames.length];
```
Please remove commented out code

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/actions/MakeStaticAction.java:
```diff
@@ -0,0 +1,153 @@
+
+	@Override
+	public void run(ITextSelection selection) {
+		if (!ActionUtil.isProcessable(fEditor))
+			return;
```
Please check all statements that require an expression block (if, while, for etc.) for having braces around the expression block. Leaving out these braces should be avoided as it is quite dangerous.

If you don't trust me, take a look at the famous Apple "goto fail" SSL bug ;-)
https://www.spiegel.de/netzwelt/web/goto-fail-apples-furchtbarer-fehler-a-955154.html

Michael5601 replied:
Done. :)

HeikoKlare reviewed (commented):
I have taken a look at all files except those directlcy related to the refactoring (i.e., within the `core.manipulation` project) again. Thanks for the significant improvements!

I have only a few minor comments left.

Commit:
improved readability

Commit:
almost working

Commit:
added general initial condition checks

Commit:
bugfix

Commit:
deleted throw

we make a daily now

Commit:
bugfix for selection of MethodInvocation before starting refactoring

Commit:
all tests working but bad code style

CodeLtDave committed:
logic fix and java docs

CodeLtDave committed:
Unified code paths for TextSelection and IMethod inputs

Unified code paths by calculating IMethod for text selection input, reducing code complexity and improving maintainability

CodeLtDave committed:
refactorings and java docs

CodeLtDave committed:
Merge branch '_DavidNEW' into _MichaelNEW

CodeLtDave committed:
merge JavaDocs and fStatus field

Commit:
small refactorings

and changing back method isOverriding to its initial state because the combination of isOverridden and isOverriding makes no sense anymore.

Commit:
small refactorings and created new class FinalConditionsChecker

Commit:
Outsourced checks

Every check from method checkFinalConditions is now checked from the class FinalConditionsChecker.

Only checks from classes MethodReferenceFinder and InstanceUsageRewriter are not outsourced yet.

Commit:
outsourced checks of MethodReferenceFinder to FinalConditionsChecker

Commit:
changed InitialConditionsChecker methods to static

Commit:
JavaDocs for InititalConditionsChecker

Commit:
fixed copyright message

Commit:
JavaDocs and copyright note for FinalConditionsChecker

Commit:
additional copyright notes
